clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +377,346 @@ impl FunctionType {
 
-        for (expected_arg, arg) in expected_args.iter().zip(func_args.iter()).into_iter() {
-            match (&expected_arg.signature, arg) {
-                (
-                    TypeSignature::TraitReferenceType(trait_id),
-                    Value::Principal(PrincipalData::Contract(contract)),
-                ) => {
-                    let contract_to_check = db
-                        .load_contract(contract)
-                        .ok_or_else(|| CheckErrors::NoSuchContract(contract.name.to_string()))?;
-                    let trait_definition = db
-                        .get_defined_trait(&trait_id.contract_identifier, &trait_id.name)
-                        .unwrap()
-                        .ok_or(CheckErrors::NoSuchContract(
-                            trait_id.contract_identifier.to_string(),
-                        ))?;
-                    contract_to_check.check_trait_compliance(trait_id, &trait_definition)?;
-                }
-                (expected_type, value) => {
-                    if !expected_type.admits(&value) {
-                        let actual_type = TypeSignature::type_of(&value);
-                        return Err(
-                            CheckErrors::TypeError(expected_type.clone(), actual_type).into()
-                        );
+        let mut arg_types = Vec::new();
+        for arg in func_args {
+            arg_types.push(self.principal_to_callable_type(arg, 1)?);
+        }
+
+        for (expected_arg, arg_type) in expected_args.iter().zip(arg_types.iter()).into_iter() {
+            inner_type_check_type(
+                db,
+                None,
+                arg_type,
+                &expected_arg.signature,
+                clarity_version,
+                1,
+                &mut LimitedCostTracker::new_free(),
+            )?;
+        }
+        Ok(returns.clone())
+    }
+}
+
+pub fn check_functions_compatible<T: CostTracker>(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    expected_sig: &FunctionSignature,
+    actual_sig: &FunctionSignature,
+    clarity_version: ClarityVersion,
+    tracker: &mut T,
+) -> bool {
+    if expected_sig.args.len() != actual_sig.args.len() {
+        return false;
+    }
+    let args_iter = expected_sig.args.iter().zip(actual_sig.args.iter());
+    for (expected_type, actual_type) in args_iter {
+        if inner_type_check_type(
+            db,
+            contract_context,
+            actual_type,
+            expected_type,
+            clarity_version,
+            1,
+            tracker,
+        )
+        .is_err()
+        {
+            return false;
+        }
+    }
+    if inner_type_check_type(
+        db,
+        contract_context,
+        &actual_sig.returns,
+        &expected_sig.returns,
+        clarity_version,
+        1,
+        tracker,
+    )
+    .is_err()
+    {
+        return false;
+    }
+    true
+}
+
+pub fn trait_check_trait_compliance<T: CostTracker>(
```

Comment:
I think there needs to be a docstring here indicating that this is Clarity2-only.  Even better, the method name itself could indicate this, so it's obvious at the call-site that this should only be invoked with Clarity2.

Reply:
Renamed to `clarity2_trait_check_trait_compliance` in 7cd1293737682ff35125f2b1069cb7cbce408ff5.